### PR TITLE
[MI-2011] Add API to listen notifications

### PR DESCRIPTION
**Pull request by ayusht2810**

### CODE REVIEW

File: server/plugin/api.go [hunk 1]
```diff
@@ -390,6 +392,34 @@ func (p *Plugin) handleGetSubscriptions(w http.ResponseWriter, r *http.Request)
...
+	attachment := &model.SlackAttachment{
+		Text: body.DetailedMessage.Markdown,
+	}
+	post := &model.Post{
+		UserId:    p.botUserID,
+		ChannelId: channelID,
+	}
+
+	model.ParseSlackAttachment(post, []*model.SlackAttachment{attachment})
+	p.API.CreatePost(post)
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:418]: this returns an error check that

File: server/plugin/api.go [hunk 2]
```diff
@@ -390,6 +392,34 @@ func (p *Plugin) handleGetSubscriptions(w http.ResponseWriter, r *http.Request)
...
+	w.Header().Add("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+}
+
 // TODO: for testing purpose, remove later
 func (p *Plugin) testAPI(w http.ResponseWriter, r *http.Request) {
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:424]: i think we can remove this now no need for this todo comment

File: server/constants/routes.go [hunk 3]
```diff
@@ -2,18 +2,18 @@ package constants
...
+	PathOAuthConnect              = "/oauth/connect"
+	PathOAuthCallback             = "/oauth/complete"
+	PathGetAllLinkedProjects      = "/project/link"
+	PathUnlinkProject             = "/project/unlink"
+	PathUser                      = "/user"
+	PathCreateTasks               = "/tasks"
+	PathLinkProject               = "/link"
+	PathGetSubscriptions          = "/subscriptions"
+	PathCreateSubscriptions       = "/subscriptions"
+	PathNotificationSubscriptions = "/notification"
```
Comment (Kshitij-Katiyar) [inline server/constants/routes.go:16]:
```suggestion
	PathSubscriptionNotifications= "/notification"
```

File: server/constants/routes.go [hunk 4]
```diff
@@ -2,18 +2,18 @@ package constants
...
+	APIPrefix                     = "/api/v1"
+	WildRoute                     = "{anything:.*}"
+	PathOAuthConnect              = "/oauth/connect"
+	PathOAuthCallback             = "/oauth/complete"
+	PathGetAllLinkedProjects      = "/project/link"
+	PathUnlinkProject             = "/project/unlink"
+	PathUser                      = "/user"
+	PathCreateTasks               = "/tasks"
+	PathLinkProject               = "/link"
+	PathGetSubscriptions          = "/subscriptions"
```
Comment (Kshitij-Katiyar) [inline server/constants/routes.go:14]: here there is no need to create 2 constants with similar values just have one constants and use it for all crud operations
```suggestion
	PathSubscriptions          = "/subscriptions"
```

File: server/plugin/api.go [hunk 5]
```diff
@@ -40,8 +41,7 @@ func (p *Plugin) InitRoutes() {
 	s.HandleFunc(constants.PathUser, p.handleAuthRequired(p.checkOAuth(p.handleGetUserAccountDetails))).Methods(http.MethodGet)
 	s.HandleFunc(constants.PathGetSubscriptions, p.handleAuthRequired(p.checkOAuth(p.handleGetSubscriptions))).Methods(http.MethodGet)
 	s.HandleFunc(constants.PathCreateSubscriptions, p.handleAuthRequired(p.checkOAuth(p.handleCreateSubscriptions))).Methods(http.MethodPost)
-	// TODO: for testing purpose, remove later
-	s.HandleFunc("/test", p.testAPI).Methods(http.MethodGet)
+	s.HandleFunc(constants.PathSubscriptionNotifications, p.handleNotificationSubscriptions).Methods(http.MethodPost)
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:44]:
```suggestion
	s.HandleFunc(constants.PathSubscriptionNotifications, p.handleSubscriptionotifications).Methods(http.MethodPost)
```

File: server/plugin/api.go [hunk 6]
```diff
@@ -390,23 +390,34 @@ func (p *Plugin) handleGetSubscriptions(w http.ResponseWriter, r *http.Request)
...
 		return
 	}
-	// TODO: for testing purposes, remove later
-	mattermostUserID := r.Header.Get(constants.HeaderMattermostUserIDAPI)
-	u, _ := p.Store.LoadUser(mattermostUserID)
-	t, _ := p.ParseAuthToken(u.AccessToken)
-	fmt.Println("\n\n\n", t)
-	res, _ := json.Marshal(response)
+
+	channelID := r.URL.Query().Get("channelID")
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:401]:
```suggestion
	if channelID := r.URL.Query().Get("channelID"); channelID== "" { 
```

Comment (ayusht2810) [reply server/plugin/api.go:401]: we are using channelID below this code.

File: server/constants/routes.go [hunk 7]
```diff
@@ -12,7 +12,7 @@ const (
 	PathCreateTasks              = "/tasks"
 	PathLinkProject              = "/link"
 	PathSubscriptions            = "/subscriptions"
-	PathSubscriptionotifications = "/notification"
+	PathSubscriptioNotifications = "/notification"
```
Comment (Kshitij-Katiyar) [inline server/constants/routes.go:15]:
```suggestion
	PathSubscriptionNotifications = "/notification"
```

File: server/plugin/api.go [hunk 8]
```diff
@@ -40,8 +41,7 @@ func (p *Plugin) InitRoutes() {
 	s.HandleFunc(constants.PathUser, p.handleAuthRequired(p.checkOAuth(p.handleGetUserAccountDetails))).Methods(http.MethodGet)
 	s.HandleFunc(constants.PathSubscriptions, p.handleAuthRequired(p.checkOAuth(p.handleGetSubscriptions))).Methods(http.MethodGet)
 	s.HandleFunc(constants.PathSubscriptions, p.handleAuthRequired(p.checkOAuth(p.handleCreateSubscriptions))).Methods(http.MethodPost)
-	// TODO: for testing purpose, remove later
-	s.HandleFunc("/test", p.testAPI).Methods(http.MethodGet)
+	s.HandleFunc(constants.PathSubscriptioNotifications, p.handleSubscriptioNotifications).Methods(http.MethodPost)
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:44]:
```suggestion
	s.HandleFunc(constants.PathSubscriptioNotifications, p.handleSubscriptionNotifications).Methods(http.MethodPost)
```

**Review by Kshitij-Katiyar — APPROVED** [review]

**Review by raghavaggarwal2308 — APPROVED** [review]

**Comment by ayusht2810** [comment]
Created new PR for these changes #36

**Pull request closed without merging**